msal/imds.py

```diff
@@ -0,0 +1,164 @@
+    """Obtains token for
+    `App Service <https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp#rest-endpoint-reference>`_
+    """
+    # Prerequisite: Create your app service https://docs.microsoft.com/en-us/azure/app-service/quickstart-python
+    # Assign it a managed identity https://docs.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp
+    # SSH into your container for testing https://docs.microsoft.com/en-us/azure/app-service/configure-linux-open-ssh-session
+    logger.debug("Obtaining token via managed identity on Azure App Service")
+    params = {
+        "api-version": "2019-08-01",
+        "resource": resource,
+        }
+    if client_id:
+        params["client_id"] = client_id
+    resp = http_client.get(
+        endpoint,
+        params=params,
+        headers={
+            "X-IDENTITY-HEADER": identity_header,
+            "Metadata": "true",  # Unnecessary yet harmless for App Service,
+            # It will be needed by Azure Automation
+            # https://docs.microsoft.com/en-us/azure/automation/enable-managed-identity-for-automation#get-access-token-for-system-assigned-managed-identity-using-http-get
+            },
+        )
+    try:
+        payload = json.loads(resp.text)
+        if payload.get("access_token") and payload.get("expires_on"):
+            return {  # Normalizing the payload into OAuth2 format
+                "access_token": payload["access_token"],
+                "expires_in": int(payload["expires_on"]) - int(time.time()),
+                "resource": payload.get("resource"),
+                "token_type": payload.get("token_type", "Bearer"),
+                }
+        return {
+            "error": "invalid_scope",  # Empirically, wrong resource ends up with a vague statusCode=500
+            "error_description": "{}, {}".format(
+                payload.get("statusCode"), payload.get("message")),
+            }
+    except ValueError:
+        logger.debug("IMDS emits unexpected payload: %s", resp.text)
+        raise
+
+
+class ManagedIdentity(object):
+    _instance, _tenant = socket.getfqdn(), "managed_identity"  # Placeholders
+
+    def __init__(self, http_client, client_id=None, token_cache=None):
+        """Create a managed identity object.
+
+        :param http_client:
+            An http client object. For example, you can use `requests.Session()`.
+
+        :param str client_id:
+            Optional.
+            It accepts the Client ID (NOT the Object ID) of your user-assigned managed identity.
+            If it is None, it means to use a system-assigned managed identity.
+
+        :param token_cache:
+            Optional. It accepts a :class:`msal.TokenCache` instance to store tokens.
+        """
+        self._http_client = http_client
+        self._client_id = client_id
+        self._token_cache = token_cache
+
+    def acquire_token(self, resource):
```

Comment:
Can we let `ManagedIdentity.acquire_token` take `scopes`, similar to `acquire_token_silent_with_error`?

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/dd75a284c1b74ecb8e9eb4d2d4598bd602b4638f/msal/application.py#L1224-L1226

Reply:
See the comment above on `_scope_to_resource()`. Partially due to the immaturity of that helper, this API only accepts resource input.

Reply:
This introduces inconsistency and Azure CLI has to make the conversion by itself.

In my opinion, managed identity should support `scope` finally.

Reply:
FWIW, in the latest commit, this was changed to `acquire_token(self, resource=None)`, to prepare for a potential future `scope` optional parameter. But, for now, the `resource` is still a required parameter.